### PR TITLE
Add fake support for pad_packed_sequence

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #185483

aten::_pad_packed_sequence was registered as CompositeImplicitAutograd even though its native kernel reads batch_sizes through data_ptr, writes the lengths output, and copies real tensor data. Fake tensor dispatch tried to run that native kernel while Dynamo was tracing pad_packed_sequence, which failed because fake tensors do not own allocated storage.

Register the operator as CompositeExplicitAutograd so fake tensor dispatch can use a Python fake implementation instead of decomposing into the native kernel. The fake implementation models the data-dependent batch dimension with an unbacked SymInt and the operator is tagged dynamic_output_shape. Since the op no longer gets implicit autograd through its native implementation, add an explicit derivative that repacks the padded output gradient with the returned lengths.

I considered handling this in torch.nn.utils.rnn or Dynamo, but that would only avoid this call site. The operator registration is the root cause, so the fix belongs at the aten/fake-tensor boundary.

Fixes #155238

Generated by my agent

Test Plan:
- ninja -C build torch_python
- ninja -C build install
- python test/dynamo/test_repros.py -k pad_packed_sequence_fake_tensor
- python test/nn/test_packed_sequence.py
- lintrunner -a

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98 @mlazos

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says the Dynamo-wrapped packed sequence failures were stale expected-failure metadata after the fake tensor fix: PackedSequenceTest.test_pack_sequence and PackedSequenceTest.test_unpack_sequence now pass under PYTORCH_TEST_WITH_DYNAMO=1. I removed those two markers, kept the remaining unrelated PackedSequence expected-failure markers, reran the targeted local tests and lintrunner, and resubmitted the ghstack PR. CI is queued on the new head 5dadf4b36e393794f4e538040116fbb38a001d17.

File: tools/autograd/derivatives.yaml [hunk 1]
```diff
@@ -2989,6 +2989,10 @@
 - name: _pack_padded_sequence(Tensor input, Tensor lengths, bool batch_first) -> (Tensor, Tensor)
   input: _pack_padded_sequence_backward_symint(grad, input.sym_sizes(), result1, batch_first)
 
+- name: _pad_packed_sequence(Tensor data, Tensor batch_sizes, bool batch_first, Scalar padding_value, int total_length) -> (Tensor, Tensor)
+  output_differentiability: [True, False]
+  data: std::get<0>(_pack_padded_sequence(grad, result1, batch_first))
```
Comment (zou3519) [inline tools/autograd/derivatives.yaml]: Is there an autograd test for this already? If there's not, then we need to add one

Comment (jansel) [reply tools/autograd/derivatives.yaml]: My agent says I added a direct autograd test for aten._pad_packed_sequence in test/nn/test_packed_sequence.py. It checks the registered backward formula against repacking the upstream gradient for both batch_first modes and total_length cases.

**Comment by jansel** [comment]
My agent says I rebased onto current main and reconciled the later #185143 workaround by removing its pad-specific Dynamo skip now that aten::_pad_packed_sequence has real fake-tensor support. I updated fullgraph coverage and preserved forward AD plus real/complex reverse and gradgrad behavior across the CompositeExplicitAutograd change. The focused Dynamo, packed-sequence, and trace-rule tests pass on rebuilt native libraries, as does lintrunner.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I addressed the follow-up review by fixing _pack_padded_sequence fake output shapes and CPU/int64 batch_sizes metadata, carrying total_length symbolically through dynamic export, adding AOTAutograd and Inductor backward coverage, registering the Inductor fallback, and updating the decomposition expectation. The Dynamo gradgrad skip remains narrowly scoped to the separate _pack_padded_sequence_backward storage-reading path; eager complex gradgrad remains covered. The full editable build, targeted CPU and CUDA checks, dynamic export test, lintrunner, and a fresh independent review all pass.

**Comment by jansel** [comment]
@claude review these changes

File: test/nn/test_packed_sequence.py [hunk 2]
```diff
@@ -509,6 +512,131 @@ def pad(tensor, length):
...
+            tangent,
+            packed.batch_sizes,
+            batch_first,
+            0.0,
+            total_length,
+        )
+
+        self.assertEqual(output, expected_output)
+        self.assertEqual(output_tangent, expected_tangent)
+        self.assertEqual(out_lengths, expected_lengths)
```
Comment (albanD) [inline test/nn/test_packed_sequence.py]: Please use the standard logic to test gradients.

**Review by albanD — CHANGES_REQUESTED** [review]

File: test/nn/test_packed_sequence.py (hunk 2 above)
Comment (jansel) [reply test/nn/test_packed_sequence.py]: My agent says I replaced the manual derivative assertions with the standard common_utils gradcheck and gradgradcheck coverage. The matrix now covers double and complex inputs, both batch layouts, exact and padded total_length values, forward AD, batched gradients, and forward-over-reverse; nn/test_packed_sequence is also in the slow-gradcheck allowlist. The standard checks exposed two additional issues, so this update adds the missing _pack_padded_sequence JVP and a shape-preserving _pad_packed_sequence VJP, with dynamic AOTAutograd/Inductor and zero-feature regression tests. The focused eager, slow-gradcheck, Dynamo, AOTAutograd, and Inductor suites all pass.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
